src/jsonpatcherproxy.js

```diff
@@ -146,6 +152,44 @@ const JSONPatcherProxy = (function() {
     instance._defaultCallback(operation);
     return reflectionResult;
   }
+  /**
+   * Test if replacing old value with new value is a significant change, i.e. whether or not
+   * it soiuld result in a patch being generated.
+   * @param {*} oldValue old value
+   * @param {*} newValue new value
+   * @param {boolean} isTreeAnArray value resides in an array
+   */
+  function isSignificantChange(oldValue, newValue, isTreeAnArray) {
+    if (isTreeAnArray) {
+      return isSignificantChangeInArray(oldValue, newValue);
+    } else {
+      return isSignificantChangeInObject(oldValue, newValue);
+    }
+  }
+  /**
+   * Test if replacing old value with new value is a significant change in an object, i.e.
+   * whether or not it should result in a patch being generated.
+   * @param {*} oldValue old value
+   * @param {*} newValue new value
+   */
+  function isSignificantChangeInObject(oldValue, newValue) {
+    return oldValue !== newValue;
+  }
+  /**
+   * Test if replacing old value with new value is a significant change in an array, i.e.
+   * whether or not it should result in a patch being generated.
+   * @param {*} oldValue old value
+   * @param {*} newValue new value
+   */
+  function isSignificantChangeInArray(oldValue, newValue) {
+    if (typeof oldValue === 'undefined') {
+      oldValue = null;
+    }
+    if (typeof newValue === 'undefined') {
+      newValue = null;
+    }
+    return oldValue !== newValue;
+  }
```

Comment:
I think this function (and tests) need to be extended cover scenario of adding a `null` beyond the end of the array:
```
arr = Array(5);
JSON.stringify(arr); // "[null,null,null,null,null]"
arr[7]; // undefined
arr[7] = null; // null - Even though `typeof oldValue === 'undefined'`, this is a significant change, as
JSON.stringify(arr); // [null,null,null,null,null,null,null,null]"
```

Reply:
@tomalec Good catch! I noticed that these cases are were not dealt with properly and had intended to file a separate issue about it. The thing is that adding `undefined` elements is broken on `master` as well and therefore i figured it was a bit out of scope for this PR.

Check out failing test on 7cf32fd09593fef2751abc51114267e4a14f3897 which is based on current `master`. Here the `undefined` element results in a `replace` patch when it should be an `add`. Based on this PR, the same test also fails, although differently: 5de895f7b922c4222e5a020b08190b67bec904f7. Here the `undefined` element is simply ignored. Note that both fail on elements which is added implicitly, without ever being set such as `arr[5]` and `arr[6]` in the example which are not present in the patch at all.

The test is based on your example with an addition:

```JS
arr = Array(5);
JSON.stringify(arr); // "[null,null,null,null,null]"
arr[7]; // undefined
arr[7] = null;
arr[8] = undefined;
// null - Even though `typeof oldValue === 'undefined'`, these are significant changes, as
JSON.stringify(arr); // [null,null,null,null,null,null,null,null,null]"
```

Reply:
See #52.

Reply:
We found another issue with this function. Is it called when you `pop` an array?
Then change from `undefined`(mapped to `null`) to not existing element.
I guess it should call `deleteTrap` but it's worth to add a test for such case as well.